File: ch10/multiserv.c
#include "tcpi.h"

int main(int argc, char *argv[])
{
        int sockfd;
        int n;
        const int on = 1;
        struct sockaddr_in sarecv;
        struct sockaddr_in safrom;
        socklen_t salen;
        char line[MAXLINE];
        char buf[INET_ADDRSTRLEN];

        if (argc != 4)
                err_quit("Usage: %s <interface> <multicast addr> <#port>", basename(argv[0]));

        if ((sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0)
                err_sys("socket error");

        if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on)) < 0)
                err_sys("setsockopt error");

        bzero(&sarecv, sizeof(struct sockaddr_in));
        sarecv.sin_family = AF_INET;
        sarecv.sin_port = htons(atoi(argv[3]));
        if (inet_pton(AF_INET, argv[2], &sarecv.sin_addr) != 1) {
                errno = EINVAL;
                err_sys("inet_pton error");
        }
        if (bind(sockfd, (struct sockaddr *)&sarecv, sizeof(struct sockaddr_in)) < 0)
                err_sys("bind error");

        if (mcast_join(sockfd, argv[1], argv[2]) < 0)
                err_sys("mcast_join error");

        for (;;) {
                salen = sizeof(struct sockaddr_in);
                n = recvfrom(sockfd, line, MAXLINE-1, 0, (struct sockaddr *)&safrom, &salen);
                if (n <= 0) break;
                line[n] = 0;
                printf("from %s: %s", inet_ntop(AF_INET, &safrom.sin_addr, buf, sizeof(buf)), line);
        }


        return 0;
}

File: ch10/multiserv.pl
#!/usr/bin/perl -w
# $Id$
use strict;
use Socket;
use IO::Handle;

my $socket;
my $addr;
my $buf = "";

if ($#ARGV != 1) {
        print STDERR "Usage: $0 <multicast ip> <port>\n";
        exit 1
}

my $mcast_addr = inet_aton($ARGV[0]);
my $port = $ARGV[1];
my $local_addr = INADDR_ANY;
my $ip_mreq = $mcast_addr . $local_addr;

socket($socket, PF_INET, SOCK_DGRAM, getprotobyname('UDP')) or die "socket: $!";

setsockopt($socket, SOL_SOCKET, SO_REUSEADDR, 1) or die "Can't reuse the socket: $!";

$addr = sockaddr_in($port, $local_addr);
bind($socket, $addr);

my $ip_level = getprotobyname('IP');
# grep -e IP_ADD_MEMBERSHIP -e IP_MULTICAST_LOOP /usr/include/netinet/in.h | grep '^#define'
my $IP_ADD_MEMBERSHIP = 12;
my $IP_MULTICAST_LOOP = 11;

setsockopt($socket, $ip_level, $IP_ADD_MEMBERSHIP, $ip_mreq) or die "Can't join group: $!";
setsockopt($socket, $ip_level, $IP_MULTICAST_LOOP, 0) or die "Can't deactive the message to loop back: $!";


while (1) {
	defined(my $recvfrom = recv($socket, $buf, 1024, 0))
		|| die "recv: $!";
        my($port, $ipaddr) = sockaddr_in($recvfrom);
        printf "from %s: %s", inet_ntoa($ipaddr), $buf;
}
